inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/cum_sum_partial_sum.cl

```diff
@@ -0,0 +1,199 @@
+    axes[1] = get_global_id(1) % OUTPUT_FEATURE_NUM;
+    axes[2] = (uint)get_global_id(2) / (OUTPUT_SIZE_Y * OUTPUT_SIZE_Z);
+    const uint zy = (uint)get_global_id(2) % (OUTPUT_SIZE_Y * OUTPUT_SIZE_Z);
+    axes[3] = zy / OUTPUT_SIZE_Y;
+    axes[4] = zy % OUTPUT_SIZE_Y;
+#endif
+}
+
+#if CUM_SUM_PARTIAL_SUM
+inline uint FUNC(get_current_index)(int axis, int i)
+{
+#ifdef REVERSE
+    return get_global_size(0)*get_local_size(0) - axis * get_local_size(0) - i - 1;
+#else
+    return axis * get_local_size(0) + i;
+#endif
+}
+
+__attribute__((intel_reqd_sub_group_size(SIMD)))
+__attribute__((reqd_work_group_size(LWS, 1, 1)))
+KERNEL(cum_sum_partial_sum)(
+    const __global INPUT0_TYPE* input,
+    __global PARTIAL_TYPE* partial)
+{
+    int axes[6], initial_axes[6];
+    FUNC_CALL(get_indices)(axes);
+    initial_axes[0] = axes[0];
+    initial_axes[1] = axes[1];
+    initial_axes[2] = axes[2];
+    initial_axes[3] = axes[3];
+    initial_axes[4] = axes[4];
+    initial_axes[5] = axes[5];
+
+    int exclusive = 0;
+#ifdef EXCLUSIVE
+#ifdef REVERSE
+    ++exclusive;
+#else
+    --exclusive;
+#endif
+#endif
+
+    INPUT0_TYPE res[BLOCK_SIZE];
+    INPUT0_TYPE prev = 0;
+    for (int i = 0; i < BLOCK_SIZE; ++i) {
+        axes[AXIS] = FUNC_CALL(get_current_index)(initial_axes[AXIS], i) + exclusive;
+        uint ind = FUNC_CALL(get_input_index)(axes[0], axes[1], axes[2], axes[3], axes[4], axes[5]);
+        res[i] = (axes[AXIS] < SUM_ITEMS_NUM && axes[AXIS] >= 0) ? input[ind] : INPUT0_VAL_ZERO;
+        res[i] += prev;
+        prev = res[i];
+    }
+
+    for (int i = 0; i < BLOCK_SIZE; ++i) {
+        axes[AXIS] = FUNC_CALL(get_current_index)(initial_axes[AXIS], i);
+        uint out_ind = FUNC_CALL(get_input_index)(axes[0], axes[1], axes[2], axes[3], axes[4], axes[5]);
+        if (axes[AXIS] < SUM_ITEMS_NUM)
+            partial[out_ind] = TO_PARTIAL_TYPE(res[i]);
+    }
+}
+#else
+inline uint FUNC(get_block_num)(int axis)
+{
+#ifdef REVERSE
+    return (SUM_ITEMS_NUM - axis - 1) / BLOCK_SIZE;
+#else
+    return axis / BLOCK_SIZE;
+#endif
+}
+
+inline uint FUNC(get_current_index)(int i)
+{
+#ifdef REVERSE
+    return SUM_ITEMS_NUM - i*BLOCK_SIZE - BLOCK_SIZE;
+#else
+    return i*BLOCK_SIZE + BLOCK_SIZE - 1;
+#endif
+}
+
+// main
+__attribute__((intel_reqd_sub_group_size(SIMD)))
+__attribute__((reqd_work_group_size(LWS, 1, 1)))
+KERNEL(cum_sum_final)(
+    const __global PARTIAL_TYPE* partial,
+    __global OUTPUT_TYPE* output)
+{
+    int axes[6];
+    FUNC_CALL(get_indices)(axes);
+    const uint batch = axes[0];
+    const uint features = axes[1];
+    const uint w = axes[2];
+    const uint z = axes[3];
+    const uint y = axes[4];
+    const uint x = axes[5];
+
+    uint ind = FUNC_CALL(get_input_index)(axes[0], axes[1], axes[2], axes[3], axes[4], axes[5]);
+    PARTIAL_TYPE res = partial[ind];
+
+    PARTIAL_TYPE sum = 0;
+    uint block_num = FUNC_CALL(get_block_num)(axes[AXIS]);
+    for (int i = 0; i < block_num; ++i) {
```

Comment:
Did you try to unroll this cycle (or to use something like i = 0; i < block_num / x; i += x) for getting better performance?

Reply:
Thank you for this idea! I tried to unroll the whole loop and it didn't give performance increase. But decreasing number of iterations (`block_num / x`) helps to get better performance!